docs/available-filters/envelope-encryption/envelope-encryption.adoc

```diff
@@ -0,0 +1,210 @@
+un-consumable (that is, the fetch will fail). In this case, the consumer offset must be advanced beyond those records.
+
+=== What exactly gets encrypted?
+
+The filter currently encrypts only *record values*.  Record keys, headers, timestamps are not encrypted.
+
+Null record values produced are passed through unencrypted.  This serves to allow tombstoning.
+
+The filter supports use-cases where some Kafka topics are configured for encryption while others are left to be
+unencrypted.  Transactions producing to both encrypted and unencrypted topics are unsupported?
+
+== How to use the filter
+
+There are three steps to using the filter.
+
+1. Setting up the KMS.
+2. Configuring the filter within Kroxylicious.
+3. Establishing the encryption key(s) within the KMS that will be used to encrypt the topics.
+
+These steps are described in the next sections.
+
+=== Setting up the KMS
+
+In order to set up the KMS provider ready to use with the filter, follow these KMS provider specific steps.
+
+include::hashicorp-vault/setup.adoc[leveloffset=3]
+
+=== Filter Configuration
+
+The filter is configured as part of the filter chain in the following way:
+
+[source, yaml]
+----
+filters:
+- type: EnvelopeEncryption                                        # <1>
+  config:
+    kms: <KMS service name>                                       # <2>
+    kmsConfig:                                                    # <3>
+      ..:
+    selector: <KEK selector service name>                         # <4>
+    selectorConfig:                                               # <5>
+      ..:
+----
+<1> The name of the filter. This must be `EnvelopeEncryption`.
+<2> The KMS service name.
+<3> Object providing configuration understood by KMS provider.
+<4> The KEK selector service name.
+<5> Object providing configuration understood by key selector.
+
+==== KMS Service configuration
+
+In order to configure the KMS Service, follow these KMS provider specific steps.
+
+include::hashicorp-vault/service.adoc[leveloffset=3]
+
+==== KEK selector configuration
+
+The role of the KEK selector is to map from the topic name to key name.  The filter looks up the resulting
+key name in the KMS.
+
+NOTE: If the filter is unable to find the key in the KMS, the filter will pass through the
+records belonging to that topic in the produce request without encrypting them.
+
+===== Template KEK Selector
+
+The `TemplateKekSelector` maps from topic name to key name.  The template understands the substitution token
+`$\{topicName}` which is replaced by the name of the topic.  It can be used to build key names
+that include the topic name being encrypted.
+
+Use the `$\{topicName}` is optional. It is possible to pass a literal string.  This will result in all topics being
+encrypted using the same key.
+
+[source, yaml]
+----
+selector: TemplateKekSelector                                 # <1>
+selectorConfig:
+  template: "topic.${topicName}"                                # <2>
```

Comment:
I don't think `topic.${topicName}` is a good example (as it would be a very odd name in the KMS) how about 
```suggestion
  template: "${topicName}_KEK"                                # <2>
```

Reply:
Good point, but as HashiCorp Vault ACL globbing only supports the wildcard in the last position, let's use a `KEK_` prefix.